hack/version_discovery.py

```diff
@@ -0,0 +1,250 @@
+                executor.submit(self.process_repository, repo, config): repo
+                for repo, config in REPOSITORIES.items()
+            }
+            for future in future_to_repo:
+                repo = future_to_repo[future]
+                try:
+                    repo_results = future.result()
+                    if repo_results:
+                        results.extend(repo_results)
+                except Exception as e:
+                    logger.error(f"Error processing {repo}: {e}")
+
+        return results
+
+    def process_repository(
+        self, repo: str, config: RepositoryConfig
+    ) -> list[RepositoryInfo]:
+        version_prefix = config.get("version_prefix", "v")
+        images = config.get("images", [])
+
+        try:
+            version_info = self.find_latest_release_version(repo, version_prefix)
+            if version_info:
+                tag_name, _ = version_info
+                return [
+                    {
+                        "repository": f"https://github.com/{repo}",
+                        "ref": tag_name,
+                        "image_url": None,
+                    }
+                ]
+        except Exception as e:
+            logger.error(f"Error finding latest release for {repo}: {e}")
+
+        if not images:
+            logger.warning(
+                f"No releases or tags found for {repo} and no images configured"
+            )
+            return []
+
+        repo_results: list[RepositoryInfo] = []
+        for image in images:
+            component_name = f"{repo}/{image.split('/')[-1]}"
+            logger.info(f"Checking {component_name} latest commit")
+
+            try:
+                commit_info = self.find_latest_commit_with_image(repo, image)
+                if commit_info:
+                    commit_sha, tag = commit_info
+                    repo_results.append(
+                        {
+                            "repository": f"https://github.com/{repo}",
+                            "ref": commit_sha,
+                            "image_url": f"{image}:{tag}",
+                        }
+                    )
+                else:
+                    logger.warning(f"No images found for {component_name}")
+            except Exception as e:
+                logger.error(f"Error finding commit with image for {repo}/{image}: {e}")
+
+        return repo_results
+
+    @staticmethod
+    def save_to_release_candidates(
+        results: list[RepositoryInfo], filename: str = "release-candidates.yaml"
+    ) -> None:
+        repo = ReleaseCandidatesRepository(filename)
+        candidate = ReleaseCandidate(components=results)
+        if repo.save(candidate):
+            logger.info(
+                f"Saved new release candidate {candidate.metadata.get('id')} to {filename}"
+            )
+        else:
+            logger.info("No changes detected or error occurred, skipping update")
+
+
+def main() -> None:
+    try:
```

Comment:
ReleaseCandidatesRepository could be initialized by factory method (so we'll load filename from env)